incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,187 @@
+		stats.objects[gk] = &object{
+			gk: gk,
+		}
+	}
+	stats.objects[gk].curReconciles.decr()
+}
+
+// WriteNamespace updates stats when writing namespace instance.
+func WriteNamespace() {
+	stats.namespaceWrites.incr()
+}
+
+// WriteHierConfig updates stats when writing hierarchyConfig instance.
+func WriteHierConfig() {
+	stats.hierConfigWrites.incr()
+}
+
+// WriteObject updates the object stats by GK when writing the object.
+func WriteObject(gk schema.GroupKind) {
+	stats.objects[gk].apiWrites.incr()
+}
+
+func initStats() {
+	objects := make(map[schema.GroupKind]*object)
+	for _, gvk := range config.GVKs {
+		objects[gvk.GroupKind()] = &object{
+			gk: gvk.GroupKind(),
+		}
+	}
+	stats = stat{
+		actionID: 1,
+		objects:  objects,
+	}
+}
+
+// StartLoggingActivity generates logs for performance testing.
+func StartLoggingActivity() {
+	initStats()
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, lastTotal, lastCur counter = 0, 0, 0
+	working := false
+	go func() {
+		// run forever
+		for {
+			// Log activity only when the controllers were still working in the last 0.5s.
+			time.Sleep(500 * time.Millisecond)
+			total = stats.totalHierConfigReconciles + getTotalObjReconciles()
+			// If lastCur is not 0 yet, still generate a log for the past 0.5s.
+			if total != lastTotal || lastCur != 0 {
+				// If the controller was previously idle, change its status and log it's started.
+				if working == false {
+					working = true
+					logActivity(log, "start")
+				}
+				logActivity(log, "continue")
+			} else {
+				// If the controller was previously working, change its status and log it's finished.
+				if working == true {
+					working = false
+					logActivity(log, "finish")
+					stats.actionID.incr()
+				}
+			}
+			lastTotal = stats.totalHierConfigReconciles + getTotalObjReconciles()
+			lastCur = stats.curHierConfigReconciles + getCurObjReconciles()
+		}
+	}()
+}
+
+func logActivity(log logr.Logger, status string) {
+	log.Info("Activity", "Action", stats.actionID, "Status", status, "HierConfigWrites", stats.hierConfigWrites, "NamespaceWrites", stats.namespaceWrites, "ObjectWrites", getObjWrites(), "TotalHierConfigReconciles", stats.totalHierConfigReconciles, "CurHierConfigReconciles", stats.curHierConfigReconciles, "TotalObjReconciles", getTotalObjReconciles(), "CurObjReconciles", getCurObjReconciles())
+}
+
+func getTotalObjReconciles() counter {
+	var total counter
+	for _, obj := range stats.objects {
+		total += obj.totalReconciles
+	}
+	return total
+}
+
+func getCurObjReconciles() counter {
+	var cur counter
+	for _, obj := range stats.objects {
+		cur += obj.curReconciles
+	}
+	return cur
+}
+
+func getObjWrites() counter {
+	var writes counter
+	for _, obj := range stats.objects {
+		writes += obj.apiWrites
+	}
+	return writes
+}
+
+func (c *counter) incr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, 1)
+	*c = counter(i)
+}
+
+func (c *counter) decr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, -1)
+	*c = counter(i)
+}
```

Comment:
Nit: I like to keep the methods of a type close to the definition of the type. I noticed you used to do that and then moved them here? You could also move these to counter.go if you don't want to clutter up the main file.